core/app/models/spree/promotion.rb

```diff
@@ -91,23 +91,28 @@ def eligible_rules(promotable, options = {})
       # Promotions without rules are eligible by default.
       return [] if rules.none?
       eligible = lambda { |r| r.eligible?(promotable, options) }
-      specific_rules = rules.for(promotable)
+      specific_rules = rules.select { |rule| rule.applicable?(promotable) }
       return [] if specific_rules.none?
 
+      eligible_rules = specific_rules.map(&eligible)
```

Comment:
I feel like this variable is misnamed. I would expect a variable named `eligible_rules` to contain `PromotionRule`s not booleans. It took me a moment to understand what was going on here.

Maybe simplify the logic below and do something like:

```Ruby
rule_eligibility = Hash[specific_rules.map do |rule|
  [rule, rule.eligible?(promotable, options)]
end]
```

Then you can dump the lambda, and the zip at the bottom.